multi-node/aws/pkg/cluster/cluster.go

```diff
@@ -42,38 +45,136 @@ func New(cfg *config.Cluster, awsDebug bool) *Cluster {
-		return "", fmt.Errorf("Invalid cloudformation stack: %v", err)
+		return "", fmt.Errorf("invalid cloudformation stack: %v", err)
 	}
 
-	return validationReport.String(), err
+	describeStacksInput := cloudformation.DescribeStacksInput{
+		StackName: aws.String(c.ClusterName),
+	}
+
+	//Find out if stack exists already. This determines whether we should do subnet conflict validation
+	var stackExists bool
+	stackNotExistExpr := regexp.MustCompile(fmt.Sprintf("^ValidationError: Stack with id %s does not exist", c.ClusterName))
+
+	describeStacksOutput, err := cfSvc.DescribeStacks(&describeStacksInput)
+	if err != nil {
+		if stackNotExistExpr.Match([]byte(err.Error())) {
+			//No results for a list operation is not an error!!! (unless your AWS)
+			stackExists = false
+		} else {
+			return "", fmt.Errorf("error describing stack: %v", err)
+		}
+	} else {
+		stackExists = true
+		if len(describeStacksOutput.Stacks) > 1 {
+			return "", fmt.Errorf("expected error: found more the one stack with name %s", c.ClusterName)
+		}
+	}
+
+	//if the stack already exists, is doesn't make sense to validate the existing VPC for subnet conflicts
+	if c.VPCID != "" && !stackExists {
+		fmt.Println("Existing VPC detected. Will validate for subnet cidr conflicts")
+		if err := c.validateExistingVPC(); err != nil {
+			return "", err
+		}
+	}
+
+	return validationReport.String(), nil
+}
+
+/*
+Calls out to AWS API to get a list of the subnets that already exist in the target existing VPC.
+Ensures that the new instance subnet being created does not overlap with any existing subnets in the VPC.
+*/
+func (c *Cluster) validateExistingVPC() error {
+	ec2Svc := ec2.New(c.session)
+
+	describeVpcsInput := ec2.DescribeVpcsInput{
+		VpcIds: []*string{aws.String(c.VPCID)},
+	}
+	vpcOutput, err := ec2Svc.DescribeVpcs(&describeVpcsInput)
+	if err != nil {
+		return fmt.Errorf("error describing existing vpc: %v", err)
+	}
+	if len(vpcOutput.Vpcs) == 0 {
+		return fmt.Errorf("could not find vpc %s in region %s", c.VPCID, c.Region)
+	}
+	if len(vpcOutput.Vpcs) > 1 {
+		//Theoretically this should never happen. If it does, we probably want to know.
+		return fmt.Errorf("found more than one vpc with id %s. this is NOT NORMAL.", c.VPCID)
+	}
+
+	existingVPC := vpcOutput.Vpcs[0]
+
+	if *existingVPC.CidrBlock != c.VPCCIDR {
+		//If this is the case, our network config validation cannot be trusted and we must abort
+		return fmt.Errorf("configured vpcCidr (%s) does not match actual existing vpc cidr (%s)", c.VPCCIDR, *existingVPC.CidrBlock)
+	}
+
+	describeSubnetsInput := ec2.DescribeSubnetsInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("vpc-id"),
+				Values: []*string{existingVPC.VpcId},
+			},
+		},
+	}
+
+	subnetOutput, err := ec2Svc.DescribeSubnets(&describeSubnetsInput)
+	if err != nil {
+		return fmt.Errorf("error describing subnets for vpc: %v", err)
+	}
+
+	//Config validation has already ensured this subnet is contained by the existing VPC CIDR
```

Comment:
This concerns me a bit. Relying on another piece of code for verification, which is merely enforced by comment+convention could lead to future issues if people are not careful to update both in unison. My initial thought would be that you either specify a VPCID or a VPC CIDR. Then validation path becomes something like:

if vpcID {
  existingVPCCidr, existingInstanceCIDRS = getCidrsFromAWS(vpcID)
}
networkValidation(existingVPCCIDR, existingInstanceCIDRs, config)

Then the validation code is more easily tested (separate concerns from network calls to aws), and is contained in one place. If no vpcID is given as part of the config, then there is no "existing" to validate against (and config passes through the same validation steps .. just with less to check against).

Doesn't have to be structured like the above, but I'm wary of relying on convention + no tests.